monitoring/nethealth.go

```diff
@@ -0,0 +1,483 @@
+	for _, packetLoss := range storedData.packetLoss {
+		if packetLoss <= packetLossThreshold {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// fetchNethealthMetrics collects the network metrics from the nethealth pod
+// specified by addr. Returns mapping of peer to networkData.
+func fetchNethealthMetrics(ctx context.Context, addr string) (map[string]networkData, error) {
+	client, err := roundtrip.NewClient(addr, "")
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to connect to nethealth service at %s.", addr)
+	}
+
+	// The two relevant metrics exposed by nethealth are 'nethealth_echo_request_total' and
+	// 'nethealth_echo_timeout_total'. We expect a pair of request/timeout metrics per peer.
+	// Example metrics received from nethealth may look something like the output below:
+	//
+	//      # HELP nethealth_echo_request_total The number of echo requests that have been sent
+	//      # TYPE nethealth_echo_request_total counter
+	//      nethealth_echo_request_total{node_name="10.128.0.96",peer_name="10.128.0.70"} 236
+	//      nethealth_echo_request_total{node_name="10.128.0.96",peer_name="10.128.0.97"} 273
+	//      # HELP nethealth_echo_timeout_total The number of echo requests that have timed out
+	//      # TYPE nethealth_echo_timeout_total counter
+	//      nethealth_echo_timeout_total{node_name="10.128.0.96",peer_name="10.128.0.70"} 37
+	//      nethealth_echo_timeout_total{node_name="10.128.0.96",peer_name="10.128.0.97"} 0
+	resp, err := client.Get(ctx, client.Endpoint("metrics"), url.Values{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var parser expfmt.TextParser
+	metricFamilies, err := parser.TextToMetricFamilies(resp.Reader())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	netData, err := parseMetrics(metricFamilies)
+	if err != nil {
+		log.WithError(err).
+			WithField("nethealth-metrics", string(resp.Bytes())).
+			Error("Received incomplete set of metrics. Could be due to a bug in nethealth or a change in labels.")
+		return nil, trace.Wrap(err)
+	}
+
+	return netData, nil
+}
+
+// parseMetrics parses the MetricsFamilies and returns the structured network
+// data. data maps a peer to its total request counter and total timeout counter.
+func parseMetrics(metricFamilies map[string]*dto.MetricFamily) (map[string]networkData, error) {
+	// echoRequests maps a peer to the current running total number of requests sent to that peer.
+	echoRequests, err := parseCounter(metricFamilies, echoRequestLabel)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to parse echo requests")
+	}
+
+	// echoTimeouts maps a peer to the current running total number of requests that timed out.
+	echoTimeouts, err := parseCounter(metricFamilies, echoTimeoutLabel)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to parse echo timeouts")
+	}
+
+	if len(echoRequests) != len(echoTimeouts) {
+		return nil, trace.BadParameter("expected equal number of counters for requests and timeouts,"+
```

Comment:
So I'm still uncertain about this error. Can you clarify how this can be resolved (the comment mentions it might be due to change of labels - does this mean invalid configuration or something else)?